packages/wasm/types/index.d.ts

```diff
@@ -9,6 +9,23 @@ import type { FilterPattern } from '@rollup/pluginutils';
  */
 export type TargetEnv = 'auto' | 'auto-inline' | 'browser' | 'node';
 
+/**
+ * The type for the plugin's loader function
+ *
+ * This is the function that ends up called when encountering a WASM import to load it and turn it into an usable object at runtime.
+ *
+ * @param {boolean} sync Whether the load should happen synchronously or not
+ * @param {string | null} filepath The path to the module.
+ * @param {string | null} src The base64-encoded source of the module
+ * @param {any} imports An object containing the module's imports
+ */
+export type WasmLoaderFunction = (
+  sync: boolean,
+  filepath: string | null,
+  src: string,
+  imports: any
+) => void;
```

Comment:
It's been a while, so I forgot a little about how it works, but as I could see, the loader function's return value isn't used anywhere? The pre-change loading code was this:
```js
`import { _loadWasmModule } from ${JSON.stringify(HELPERS_ID)};
export default function(imports){return _loadWasmModule(${+isSync}, ${publicFilepath}, ${src}, imports)}`
```

but as far as I can see, that return could just as well not be here, since I don't see how you could make use of it and there's no higher-level code expecting anything out of it…
So at that point, if you want the loader to do clever things with the compiled module/instance, just provide a custom loader that does so since there's not much we could do over that that wouldn't lock in details about what's expected of the loader.

Reply:
> I don't see how you could make use of it and there's no higher-level code expecting anything out of it

The code importing the wasm module expects the return value.

If the `return` is removed from the following example, downstream libraries will fail

```js
{
  loader: async (sync, filepath, src, imports) => {
    const fs = require("fs/promises");
    const path = require("path");
    const wasmBuffer = await fs.readFile(path.resolve(__dirname, filepath));
    return WebAssembly.compile(wasmBuffer);
  },
}
```

While it is possible to write a wasm library with a loader that assigns to a global variable instead of use the return value, I don't think that would be recommended.